codec-http/src/main/java/io/netty/handler/codec/http/HttpContentEncoder.java

```diff
@@ -289,34 +289,34 @@ private boolean encodeContent(HttpContent c, List<Object> out) {
 
     @Override
     public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
-        cleanup();
+        cleanupSafely(ctx);
         super.handlerRemoved(ctx);
     }
 
     @Override
     public void channelInactive(ChannelHandlerContext ctx) throws Exception {
-        cleanup();
+        cleanupSafely(ctx);
         super.channelInactive(ctx);
     }
 
     private void cleanup() {
         if (encoder != null) {
             // Clean-up the previous encoder if not cleaned up correctly.
-            if (encoder.finish()) {
-                for (;;) {
-                    ByteBuf buf = encoder.readOutbound();
-                    if (buf == null) {
-                        break;
-                    }
-                    // Release the buffer
-                    // https://github.com/netty/netty/issues/1524
-                    buf.release();
-                }
-            }
+            encoder.finishAndReleaseAll();
             encoder = null;
         }
     }
 
+    private void cleanupSafely(ChannelHandlerContext ctx) {
+        try {
+            cleanup();
+        } catch (Exception e) {
```

Comment:
should this be `Throwable` like in `AbstractChannelHandlerContext` so nothing escapes the pipeline?

Reply:
@Scottmitch I used `Exception` as we want to review our usage of `catch (Throwable...)` anyway as most likely we should do this most of the times. That said I am not strong on this.

Reply:
I see ... this may fall into the "we won't cleanup correctly if we let Throwable slip through" category ... because the pipeline won't be notified of events.

Reply:
True enough... let me just use `Throwable` for now

Reply:
Or keep exception and use try finally?